libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/navigation/NavigationStateViewModel.kt

```diff
@@ -0,0 +1,119 @@
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+/**
+ * Class that manages NavigationState as following:
+ *
+ * ```
+ *  [ FreeDrive ]━━( destination != null )━━━━━━>>[ RoutePreview ]<<━━━━━━━━━━━━┓
+ *       ^         ( & routes not empty  )               ┃                      ┃
+ *       ┃                                  (activeNavigationStarted=true)      ┃
+ *       ┃                                               ┃                      ┃
+ *       ┃                                               ┃         (activeNavigationStarted=false)
+ *       ┣━━━━( destination == null )━━━━┓               v                      ┃
+ *       ┗━━━━( routes empty        )━━━━╋━━━━━[ ActiveNavigation ]━━━━━━━━━━━━━┫
+ *                                       ┃               ┃                      ┃
+ *                                       ┃  (onFinalDestinationArrival())       ┃
+ *                                       ┃               ┃                      ┃
+ *                                       ┃               v                      ┃
+ *                                       ┗━━━━━━━━━━[ Arrival ]━━━━━━━━━━━━━━━━━┛
+ * ```
+ */
+internal class NavigationStateViewModel(
+    mutableState: MutableStateFlow<NavigationState>,
+    private val routesState: StateFlow<RoutesState>
+) : UIViewModel<NavigationState, NavigationStateAction>(mutableState) {
+
+    override fun process(
+        mapboxNavigation: MapboxNavigation,
+        state: NavigationState,
+        action: NavigationStateAction
+    ): NavigationState {
+        return when (action) {
+            is NavigationStateAction.Update -> action.state
+        }
+    }
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        mainJobControl.scope.launch {
+            routesState.collect {
+                val state = getNavigationState(
+                    it.destination,
+                    it.navigationStarted,
+                    mapboxNavigation.getRoutes()
+                )
+                updateState(state)
+            }
+        }
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowRoutesUpdated().collect {
+                val state = getNavigationState(
+                    routesState.value.destination,
+                    routesState.value.navigationStarted,
+                    it.routes
+                )
+                updateState(state)
+            }
+        }
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowOnFinalDestinationArrival().collect {
+                val inActiveNav =
+                    state.value == NavigationState.ActiveNavigation
+                val arrived = it.currentState == RouteProgressState.COMPLETE
+                if (inActiveNav && arrived) {
+                    updateState(NavigationState.Arrival)
+                }
+            }
+        }
```

Comment:
We need to establish a definition for what the mapping is for each `NavigationState`

It will allow us to delete the dependencies for `NavigationStateViewModel` and the circular dependency created with `RoutesViewModel`

The definition for "why exactly navigation state" needs RoutesViewModel will help us decide how to organize it.

After digging, the reason we cannot reduce the NavigationState without `RoutesViewModel` is because we have a RoutePreview state. The RoutePreview state is not supported inside `MapboxNavigation` because, the moment you set a route, you start ActiveNavigation.

That's the root issue, solution appears ambiguous on how exactly we handle it in drop-in. But we do have "preferable" solutions. And we prefer to remove the circular dependency

Reply:
Here is an idea for removing the circular dependency.

Remove all the logic from the NavigationStateViewModel. Add a NavigationStateComponent which depends on the RoutesViewModel and the NavigationStateViewModel. It then decides how to update the state 

https://github.com/mapbox/mapbox-navigation-android/compare/tr-%231438-dropin-info-panel...km-info-panel-refactor?expand=1

Reply:
I already had this with a different name (NavigationStateManager) and references to each individual Flow instead of ViewModels.
https://github.com/mapbox/mapbox-navigation-android/pull/5506/commits/81d4ab1d6998b724c34f9609813f7d30a0eb870a#diff-bb50b0e1f2fbbb2d2ccf984bceb5baaa41496d5f6ac1818924c68d9ec7c09e93

I'll merge your branch to mine and take it from there.